multiLevelCoSurrogates/CandidateArchiveNew.py

```diff
@@ -0,0 +1,181 @@
+        return archive
+
+
+    @property
+    def fidelities(self):
+        return self._all_fidelities.keys()
+
+
+    def addcandidates(self, candidates, fitnesses, fidelity=None):
+        """Add candidate, fitness pairs to the archive for given fidelity.
+        Will overwrite fitness value if already present.
+        """
+        for candidate, fitness in zip(candidates, np.ravel(fitnesses)):
+            self.addcandidate(candidate, fitness, fidelity)
+
+
+    def addcandidate(self, candidate, fitness, fidelity=None):
+        """Add a candidate, fitness pair to the archive for given fidelity.
+        Will overwrite fitness value if already present
+        """
+        # unspecified fidelity when explicit fidelities are present raises Error
+        if fidelity is None and len(self.fidelities) >= 1 and fidelity not in self.fidelities:
+            raise ValueError(f'Since explcit fidelities are present, new candidates '
+                             f'cannot be added with implicit fidelity. Fidelities '
+                             f'currently present: {self.fidelities}')
+
+        try:
+            idx = self.candidates.index(candidate)
+            self.candidates[idx].fidelities[fidelity] = fitness
+
+        except ValueError:
+            # candidates is not yet present
+            new_idx = len(self.candidates)
+            fidelities = {fidelity: fitness}
+            self.candidates.append(Candidate(new_idx, candidate, fidelities))
+
+        # create key entry if it does not yet exist
+        self._all_fidelities[fidelity] = None
+        self._updateminmax(fitness, fidelity)
+
+
+    def getfitnesses(self, candidates: Iterable, fidelity: Union[str, Iterable[str]]) -> Iterable:
+        """Return the relevant fitness values for the given candidates"""
+
+        if isinstance(fidelity, str):
+            fidelity = [fidelity]
+
+        # retrieve all candidates by index, fails if candidate is not in archive
+        candidate_indices = [
+            self.candidates.index(candidate)
+            for candidate in candidates
+        ]
+
+        fitnesses = np.array([
+            [self.candidates[i].fidelities.get(f, np.nan) for f in fidelity]
+            for i in candidate_indices
+        ])
+
+        if len(fidelity) == 1:
+            return fitnesses.reshape(-1)
+        return fitnesses
+
+
+    def getcandidates(self, fidelity=None):
+        """Retrieve candidates and fitnesses from the archive.
+        :fidelity:  (List of) fidelities to select by. Default: all
+        :return:    Candidates, Fitnesses (tuple of numpy arrays)
+        """
+        if not fidelity:
+            fidelity = self.fidelities
+        elif isinstance(fidelity, str):
+            fidelity = [fidelity]
+
+        candidates, fitnesses = [], []
+        for candidate in self.candidates:
+            fitness = [
+                candidate.fidelities.get(fid, np.nan)
+                for fid in fidelity
+            ]
```

Comment:
Because I need the intermediate value `fitness` to check whether to add an entry for both `fitnesses` and `candidates`. In your comprehension, that intermediate value wouldn't exist ;)